In [1]:
import pandas as pd
import tensorflow as tf
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
# import pandas_profiling 
import seaborn as sns
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
from sklearn.model_selection import train_test_split
%matplotlib inline

In [59]:
data=pd.read_csv("paypoint_data.csv")

In [4]:
data

TransactionId         BatchId       AccountId  \
0       TransactionId_76872   BatchId_16131  AccountId_2957   
1       TransactionId_73770   BatchId_15643  AccountId_4841   
2       TransactionId_26203   BatchId_51941  AccountId_4119   
3         TransactionId_380  BatchId_103161   AccountId_648   
4       TransactionId_28295   BatchId_18780  AccountId_4841   
5       TransactionId_23223   BatchId_35954  AccountId_1078   
6      TransactionId_228063  BatchId_118460  AccountId_1441   
7      TransactionId_200640   BatchId_18561  AccountId_4841   
8       TransactionId_52905   BatchId_91774   AccountId_171   
9      TransactionId_230262   BatchId_83409   AccountId_710   
10      TransactionId_52800  BatchId_113388  AccountId_1624   
11      TransactionId_33857  BatchId_136194  AccountId_4841   
12      TransactionId_94963  BatchId_104736   AccountId_710   
13     TransactionId_204687   BatchId_59411   AccountId_710   
14      TransactionId_86903   BatchId_59411   AccountId_710   
15     TransactionId_228832   BatchId_59411   AccountId_710   
16     TransactionId_222975   BatchId_11113  AccountId_1685   
17      TransactionId_98367    BatchId_6789  AccountId_1685   
18      TransactionId_64044  BatchId_117711  AccountId_4841   
19     TransactionId_229822   BatchId_89644  AccountId_2081   
20      TransactionId_94363   BatchId_94913  AccountId_4841   
21      TransactionId_93543  BatchId_101537  AccountId_2081   
22     TransactionId_238239    BatchId_8114  AccountId_4841   
23      TransactionId_79885   BatchId_48769  AccountId_1815   
24     TransactionId_227232   BatchId_50160  AccountId_1214   
25      TransactionId_74649  BatchId_100366  AccountId_4841   
26       TransactionId_7396  BatchId_134311  AccountId_1815   
27     TransactionId_226754   BatchId_60308  AccountId_1616   
28     TransactionId_236360   BatchId_48611  AccountId_4841   
29      TransactionId_89705   BatchId_16113  AccountId_2455   
...                     ...             ...             ...   
95632   TransactionId_70563    BatchId_3811   AccountId_527   
95633   TransactionId_66526   BatchId_98969  AccountId_2014   
95634    TransactionId_5394  BatchId_131781  AccountId_1965   
95635   TransactionId_22523   BatchId_35909  AccountId_1585   
95636  TransactionId_222982  BatchId_108518  AccountId_4841   
95637   TransactionId_53307  BatchId_116096  AccountId_4841   
95638   TransactionId_20690   BatchId_18591  AccountId_1121   
95639   TransactionId_73868   BatchId_94787  AccountId_4216   
95640  TransactionId_232580   BatchId_16309  AccountId_1119   
95641  TransactionId_234543   BatchId_88196  AccountId_4216   
95642  TransactionId_227975   BatchId_55755  AccountId_4841   
95643  TransactionId_224220   BatchId_89178  AccountId_1218   
95644   TransactionId_49829  BatchId_137156  AccountId_2447   
95645  TransactionId_232922   BatchId_69178  AccountId_1716   
95646  TransactionId_228602   BatchId_88011  AccountId_4841   
95647   TransactionId_74925   BatchId_39059  AccountId_4841   
95648   TransactionId_96253   BatchId_34111  AccountId_1685   
95649  TransactionId_238376  BatchId_107100   AccountId_115   
95650   TransactionId_36354   BatchId_81314   AccountId_115   
95651   TransactionId_26042   BatchId_49101   AccountId_115   
95652   TransactionId_72994   BatchId_66957   AccountId_610   
95653   TransactionId_68852  BatchId_100900  AccountId_4841   
95654   TransactionId_28230   BatchId_30110   AccountId_610   
95655   TransactionId_38907   BatchId_93168  AccountId_4841   
95656   TransactionId_68402   BatchId_17093  AccountId_1659   
95657   TransactionId_89882   BatchId_96668  AccountId_4841   
95658   TransactionId_92597    BatchId_1501  AccountId_2429   
95659   TransactionId_82502  BatchId_118603  AccountId_4841   
95660  TransactionId_236354   BatchId_70934  AccountId_1246   
95661   TransactionId_35670   BatchId_39117  AccountId_4841   

            SubscriptionId       CustomerId CurrencyCode  CountryCode  \
0       SubscriptionId_887  Custom

In [5]:
data["FraudResult"].value_counts()

0    95469
1      193
Name: FraudResult, dtype: int64

In [6]:
data.drop("TransactionStartTime",axis=1,inplace=True)

In [7]:
data.head()

TransactionId         BatchId       AccountId       SubscriptionId  \
0  TransactionId_76872   BatchId_16131  AccountId_2957   SubscriptionId_887   
1  TransactionId_73770   BatchId_15643  AccountId_4841  SubscriptionId_3829   
2  TransactionId_26203   BatchId_51941  AccountId_4119   SubscriptionId_222   
3    TransactionId_380  BatchId_103161   AccountId_648  SubscriptionId_2185   
4  TransactionId_28295   BatchId_18780  AccountId_4841  SubscriptionId_3829   

        CustomerId CurrencyCode  CountryCode    ProviderId     ProductId  \
0  CustomerId_4406          ZWL          263  ProviderId_6  ProductId_10   
1  CustomerId_4406          ZWL          263  ProviderId_4   ProductId_6   
2  CustomerId_4683          ZWL          263  ProviderId_6   ProductId_1   
3   CustomerId_988          ZWL          263  ProviderId_1  ProductId_21   
4   CustomerId_988          ZWL          263  ProviderId_4   ProductId_6   

      ProductCategory    ChannelId   Amount  Value  PricingStrategy  \
0             airtime  ChannelId_3   1000.0   1000                2   
1  financial_services  ChannelId_2    -20.0     20                2   
2             airtime  ChannelId_3    500.0    500                2   
3        utility_bill  ChannelId_3  20000.0  21800                2   
4  financial_services  ChannelId_2   -644.0    644                2   

   FraudResult  
0            0  
1            0  
2            0  
3            0  
4            0

In [8]:
data.drop("TransactionId",axis=1,inplace=True)

In [9]:
data.head()

BatchId       AccountId       SubscriptionId       CustomerId  \
0   BatchId_16131  AccountId_2957   SubscriptionId_887  CustomerId_4406   
1   BatchId_15643  AccountId_4841  SubscriptionId_3829  CustomerId_4406   
2   BatchId_51941  AccountId_4119   SubscriptionId_222  CustomerId_4683   
3  BatchId_103161   AccountId_648  SubscriptionId_2185   CustomerId_988   
4   BatchId_18780  AccountId_4841  SubscriptionId_3829   CustomerId_988   

  CurrencyCode  CountryCode    ProviderId     ProductId     ProductCategory  \
0          ZWL          263  ProviderId_6  ProductId_10             airtime   
1          ZWL          263  ProviderId_4   ProductId_6  financial_services   
2          ZWL          263  ProviderId_6   ProductId_1             airtime   
3          ZWL          263  ProviderId_1  ProductId_21        utility_bill   
4          ZWL          263  ProviderId_4   ProductId_6  financial_services   

     ChannelId   Amount  Value  PricingStrategy  FraudResult  
0  ChannelId_3   1000.0   1000                2            0  
1  ChannelId_2    -20.0     20                2            0  
2  ChannelId_3    500.0    500                2            0  
3  ChannelId_3  20000.0  21800                2            0  
4  ChannelId_2   -644.0    644                2            0

In [11]:
data.nunique()

BatchId            62286
AccountId           2355
SubscriptionId      3627
CustomerId          3742
CurrencyCode           1
CountryCode            1
ProviderId             6
ProductId             23
ProductCategory        9
ChannelId              4
Amount              1676
Value               1517
PricingStrategy        4
FraudResult            2
dtype: int64

To calculate the memory the data set occupy

In [13]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 95662 entries, 0 to 95661
Data columns (total 14 columns):
BatchId            95662 non-null object
AccountId          95662 non-null object
SubscriptionId     95662 non-null object
CustomerId         95662 non-null object
CurrencyCode       95662 non-null object
CountryCode        95662 non-null int64
ProviderId         95662 non-null object
ProductId          95662 non-null object
ProductCategory    95662 non-null object
ChannelId          95662 non-null object
Amount             95662 non-null float64
Value              95662 non-null int64
PricingStrategy    95662 non-null int64
FraudResult        95662 non-null int64
dtypes: float64(1), int64(4), object(9)
memory usage: 10.2+ MB


In [14]:
data.drop(["CurrencyCode","CountryCode"],axis=1,inplace=True)

In [15]:
data.head()

BatchId       AccountId       SubscriptionId       CustomerId  \
0   BatchId_16131  AccountId_2957   SubscriptionId_887  CustomerId_4406   
1   BatchId_15643  AccountId_4841  SubscriptionId_3829  CustomerId_4406   
2   BatchId_51941  AccountId_4119   SubscriptionId_222  CustomerId_4683   
3  BatchId_103161   AccountId_648  SubscriptionId_2185   CustomerId_988   
4   BatchId_18780  AccountId_4841  SubscriptionId_3829   CustomerId_988   

     ProviderId     ProductId     ProductCategory    ChannelId   Amount  \
0  ProviderId_6  ProductId_10             airtime  ChannelId_3   1000.0   
1  ProviderId_4   ProductId_6  financial_services  ChannelId_2    -20.0   
2  ProviderId_6   ProductId_1             airtime  ChannelId_3    500.0   
3  ProviderId_1  ProductId_21        utility_bill  ChannelId_3  20000.0   
4  ProviderId_4   ProductId_6  financial_services  ChannelId_2   -644.0   

   Value  PricingStrategy  FraudResult  
0   1000                2            0  
1     20                2            0  
2    500                2            0  
3  21800                2            0  
4    644                2            0

# converting textual data to numeric / text to numbers, skleen is an instance of a class then label encorder are blue prints 

In [ ]:
from sklearn.preprocessing import LabelEncoder
encoder=LabelEncoder()

In [16]:
data.head()

BatchId       AccountId       SubscriptionId       CustomerId  \
0   BatchId_16131  AccountId_2957   SubscriptionId_887  CustomerId_4406   
1   BatchId_15643  AccountId_4841  SubscriptionId_3829  CustomerId_4406   
2   BatchId_51941  AccountId_4119   SubscriptionId_222  CustomerId_4683   
3  BatchId_103161   AccountId_648  SubscriptionId_2185   CustomerId_988   
4   BatchId_18780  AccountId_4841  SubscriptionId_3829   CustomerId_988   

     ProviderId     ProductId     ProductCategory    ChannelId   Amount  \
0  ProviderId_6  ProductId_10             airtime  ChannelId_3   1000.0   
1  ProviderId_4   ProductId_6  financial_services  ChannelId_2    -20.0   
2  ProviderId_6   ProductId_1             airtime  ChannelId_3    500.0   
3  ProviderId_1  ProductId_21        utility_bill  ChannelId_3  20000.0   
4  ProviderId_4   ProductId_6  financial_services  ChannelId_2   -644.0   

   Value  PricingStrategy  FraudResult  
0   1000                2            0  
1     20                2            0  
2    500                2            0  
3  21800                2            0  
4    644                2            0

# To select data with same structer here is the code

In [18]:
data.select_dtypes("object")

BatchId       AccountId       SubscriptionId       CustomerId  \
0       BatchId_16131  AccountId_2957   SubscriptionId_887  CustomerId_4406   
1       BatchId_15643  AccountId_4841  SubscriptionId_3829  CustomerId_4406   
2       BatchId_51941  AccountId_4119   SubscriptionId_222  CustomerId_4683   
3      BatchId_103161   AccountId_648  SubscriptionId_2185   CustomerId_988   
4       BatchId_18780  AccountId_4841  SubscriptionId_3829   CustomerId_988   
5       BatchId_35954  AccountId_1078  SubscriptionId_4238  CustomerId_1432   
6      BatchId_118460  AccountId_1441  SubscriptionId_1980  CustomerId_2858   
7       BatchId_18561  AccountId_4841  SubscriptionId_3829  CustomerId_2858   
8       BatchId_91774   AccountId_171  SubscriptionId_4731   CustomerId_598   
9       BatchId_83409   AccountId_710   SubscriptionId_920  CustomerId_1053   
10     BatchId_113388  AccountId_1624  SubscriptionId_3511  CustomerId_3052   
11     BatchId_136194  AccountId_4841  SubscriptionId_3829  CustomerId_3052   
12     BatchId_104736   AccountId_710   SubscriptionId_920  CustomerId_1053   
13      BatchId_59411   AccountId_710   SubscriptionId_920  CustomerId_1053   
14      BatchId_59411   AccountId_710   SubscriptionId_920  CustomerId_1053   
15      BatchId_59411   AccountId_710   SubscriptionId_920  CustomerId_1053   
16      BatchId_11113  AccountId_1685  SubscriptionId_4626  CustomerId_3105   
17       BatchId_6789  AccountId_1685  SubscriptionId_4626  CustomerId_3105   
18     BatchId_117711  AccountId_4841  SubscriptionId_3829  CustomerId_3105   
19      BatchId_89644  AccountId_2081  SubscriptionId_4312  CustomerId_3507   
20      BatchId_94913  AccountId_4841  SubscriptionId_3829  CustomerId_3507   
21     BatchId_101537  AccountId_2081  SubscriptionId_4312  CustomerId_3507   
22       BatchId_8114  AccountId_4841  SubscriptionId_3829  CustomerId_3507   
23      BatchId_48769  AccountId_1815  SubscriptionId_4080  CustomerId_2216   
24      BatchId_50160  AccountId_1214  SubscriptionId_4722  CustomerId_2728   
25     BatchId_100366  AccountId_4841  SubscriptionId_3829  CustomerId_2728   
26     BatchId_134311  AccountId_1815  SubscriptionId_4080  CustomerId_2216   
27      BatchId_60308  AccountId_1616  SubscriptionId_4410  CustomerId_1995   
28      BatchId_48611  AccountId_4841  SubscriptionId_3829  CustomerId_1995   
29      BatchId_16113  AccountId_2455   SubscriptionId_946  CustomerId_3890   
...               ...             ...                  ...              ...   
95632    BatchId_3811   AccountId_527  SubscriptionId_3794   CustomerId_872   
95633   BatchId_98969  AccountId_2014  SubscriptionId_1575  CustomerId_3439   
95634  BatchId_131781  AccountId_1965  SubscriptionId_4144  CustomerId_3390   
95635   BatchId_35909  AccountId_1585  SubscriptionId_2843  CustomerId_3002   
95636  BatchId_108518  AccountId_4841  SubscriptionId_3829  CustomerId_3390   
95637  BatchId_116096  AccountId_4841  SubscriptionId_3829  CustomerId_3002   
95638   BatchId_18591  AccountId_1121  SubscriptionId_1780  CustomerId_1593   
95639   BatchId_94787  AccountId_4216  SubscriptionId_2193  CustomerId_4782   
95640   BatchId_16309  AccountId_1119  SubscriptionId_4187  CustomerId_1579   
95641   BatchId_88196  AccountId_4216  SubscriptionId_2193  CustomerId_4782   
95642   BatchId_55755  AccountId_4841  SubscriptionId_3829  CustomerId_4782   
95643   BatchId_89178  AccountId_1218  SubscriptionId_1104  CustomerId_1681   
95644  BatchId_137156  AccountId_2447  SubscriptionId_4352  CustomerId_3882   
95645   BatchId_69178  AccountId_1716  SubscriptionId_3565  CustomerId_2111   
95646   BatchId_88011  AccountId_4841  SubscriptionId_3829  CustomerId_3882   
95647   BatchId_39059  AccountId_4841  SubscriptionId_3829  CustomerId_2111   
95648   BatchId_34111  AccountId_1685    SubscriptionId_84  CustomerId_2067   
95649  BatchId_107100   AccountId_115   SubscriptionId_137   CustomerId_539   
95650   BatchId_81314   AccountId_115   SubscriptionId_137   Cus

# We are fitting and transforming number more like rearranging eg ABC to 123 so A will be 1 b 2

In [21]:
from sklearn.preprocessing import LabelEncoder
encoder=LabelEncoder()

In [23]:
for feature_name in data.select_dtypes("object"):
 data[feature_name]=encoder.fit_transform(data[feature_name])

In [24]:
data.head()

BatchId  AccountId  SubscriptionId  CustomerId  ProviderId  ProductId  \
0    20131       1413            3535        2584           5          1   
1    19722       1999            2366        2584           3         19   
2    35568       1544             996        2806           5          0   
3     1498       2112             974        3733           0         11   
4    22088       1999            2366        3733           3         19   

   ProductCategory  ChannelId   Amount  Value  PricingStrategy  FraudResult  
0                0          2   1000.0   1000                2            0  
1                2          1    -20.0     20                2            0  
2                0          2    500.0    500                2            0  
3                8          2  20000.0  21800                2            0  
4                2          1   -644.0    644                2            0

# Now scale the data by calculating mean and standard deciation its standard scaling.

In [25]:
from sklearn.preprocessing import StandardScaler
scaler=StandardScaler()

# Fit we are calculating mean and transforming is Sd

In [27]:
target=data["FraudResult"]
independent=data.drop("FraudResult",axis=1)

In [28]:
scaled=scaler.fit_transform(independent)

C:\ProgramData\Anaconda3\lib\site-packages\sklearn\preprocessing\data.py:645: DataConversionWarning: Data with input dtype int32, int64, float64 were all converted to float64 by StandardScaler.
  return self.partial_fit(X, y)
C:\ProgramData\Anaconda3\lib\site-packages\sklearn\base.py:464: DataConversionWarning: Data with input dtype int32, int64, float64 were all converted to float64 by StandardScaler.
  return self.fit(X, **fit_params).transform(X)


In [33]:
independent.head()

BatchId  AccountId  SubscriptionId  CustomerId  ProviderId  ProductId  \
0    20131       1413            3535        2584           5          1   
1    19722       1999            2366        2584           3         19   
2    35568       1544             996        2806           5          0   
3     1498       2112             974        3733           0         11   
4    22088       1999            2366        3733           3         19   

   ProductCategory  ChannelId   Amount  Value  PricingStrategy  
0                0          2   1000.0   1000                2  
1                2          1    -20.0     20                2  
2                0          2    500.0    500                2  
3                8          2  20000.0  21800                2  
4                2          1   -644.0    644                2

In [29]:
scaled

array([[-0.47222562,  0.08999884,  1.6686459 , ..., -0.04637114,
        -0.07229109, -0.34925239],
       [-0.49482353,  0.91520895,  0.34644914, ..., -0.05464323,
        -0.08025072, -0.34925239],
       [ 0.38069336,  0.27447414, -1.20308855, ..., -0.05042608,
        -0.07635213, -0.34925239],
       ...,
       [-0.93899116,  0.91520895,  0.34644914, ..., -0.05464323,
        -0.08025072, -0.34925239],
       [ 0.96182971, -1.60266763,  1.45261547, ..., -0.03015135,
        -0.05604697, -0.34925239],
       [-0.03164935,  0.91520895,  0.34644914, ..., -0.05496763,
        -0.07992583, -0.34925239]])

# To display information above(scaled data) in a table write the code below

In [31]:
x=pd.DataFrame(scaled,columns=independent.columns)

In [32]:
x.head()

BatchId  AccountId  SubscriptionId  CustomerId  ProviderId  ProductId  \
0 -0.472226   0.089999        1.668646    0.595070    1.025848  -1.578054   
1 -0.494824   0.915209        0.346449    0.595070   -0.502027   0.916063   
2  0.380693   0.274474       -1.203089    0.807169    1.025848  -1.716616   
3 -1.501729   1.074336       -1.227972    1.692827   -2.793841  -0.192433   
4 -0.364098   0.915209        0.346449    1.692827   -0.502027   0.916063   

   ProductCategory  ChannelId    Amount     Value  PricingStrategy  
0        -0.799047   0.746738 -0.046371 -0.072291        -0.349252  
1         0.491064  -1.175144 -0.054643 -0.080251        -0.349252  
2        -0.799047   0.746738 -0.050426 -0.076352        -0.349252  
3         4.361398   0.746738  0.107717  0.096648        -0.349252  
4         0.491064  -1.175144 -0.059704 -0.075183        -0.349252

# Training split test { learn with 80% of data and excute what learnt with the remaining data} 

In [34]:
from sklearn.model_selection import train_test_split

In [36]:
y=target

In [37]:
x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=0.2,random_state=100)

# telling/training the system how to allocate data, ie we are teaching the machine what to do with data

In [45]:
from sklearn.naive_bayes import BernoulliNB
model=BernoulliNB()

In [46]:
model.fit(x_train,y_train)

BernoulliNB(alpha=1.0, binarize=0.0, class_prior=None, fit_prior=True)

# Model Evaluation 

# 1. Testing Accuracy as a percentage

In [48]:
accuracy=model.score(x_test,y_test)

In [49]:
accuracy*100

98.97559190926671

# 2. Confusion Matrics

In [50]:
from sklearn.metrics import confusion_matrix,classification_report

In [51]:
predictions=model.predict(x_test)

In [53]:
print(confusion_matrix(y_test,predictions))

[[18926   162]
 [   34    11]]


In [55]:
print(classification_report(y_test,predictions))

              precision    recall  f1-score   support

           0       1.00      0.99      0.99     19088
           1       0.06      0.24      0.10        45

   micro avg       0.99      0.99      0.99     19133
   macro avg       0.53      0.62      0.55     19133
weighted avg       1.00      0.99      0.99     19133



# Precission= TP/(TP+FP)
#                   = 11/(11+162)


# Recall       = TP/(TP+FN)

#                   =11/(11+


# F1 Score= 2xPxR/(P+R)
#                 =


# Support = All values in a test set
#                =11+34
#                =45


# How to to Pandas Profiling

In [57]:
data.describe()

BatchId     AccountId  SubscriptionId    CustomerId    ProviderId  \
count  95662.000000  95662.000000    95662.000000  95662.000000  95662.000000   
mean   28677.822249   1349.089827     2059.692302   1961.152067      3.657157   
std    18099.117316    710.125889      884.139292   1046.685310      1.309014   
min        0.000000      0.000000        0.000000      0.000000      0.000000   
25%    12002.000000    710.000000     1538.000000   1131.000000      3.000000   
50%    26621.500000   1563.000000     2366.000000   1969.000000      4.000000   
75%    44510.750000   1999.000000     2467.000000   2891.000000      5.000000   
max    62285.000000   2354.000000     3626.000000   3741.000000      5.000000   

          ProductId  ProductCategory     ChannelId        Amount  \
count  95662.000000     95662.000000  95662.000000  9.566200e+04   
mean      12.388786         1.238726      1.611455  6.717846e+03   
std        7.217021         1.550262      0.520326  1.233068e+05   
min        0.000000         0.000000      0.000000 -1.000000e+06   
25%        6.000000         0.000000      1.000000 -5.000000e+01   
50%       16.000000         2.000000      2.000000  1.000000e+03   
75%       19.000000         2.000000      2.000000  2.800000e+03   
max       22.000000         8.000000      3.000000  9.880000e+06   

              Value  PricingStrategy   FraudResult  
count  9.566200e+04     95662.000000  95662.000000  
mean   9.900584e+03         2.255974      0.002018  
std    1.231221e+05         0.732924      0.044872  
min    2.000000e+00         0.000000      0.000000  
25%    2.750000e+02         2.000000      0.000000  
50%    1.000000e+03         2.000000      0.000000  
75%    5.000000e+03         2.000000      0.000000  
max    9.880000e+06         4.000000      1.000000